# Ejercicios en Python
##### por Andrés Calderón Bernal

## 1. Práctica Uno: Programas Secuenciales
Desarrollar un programa que permita registrar las notas de una asignatura X, de uno o varios estudiantes.

El programa debe solicitar, inicialmente, la cantidad de estudiantes a registrar y, para cada uno de ellos, pedir el nombre y apellido del estudiante, la cantidad de notas a ingresar (con un mínimo de tres notas) y las notas correspondientes, las cuales deben estar en un rango entre 1.0 y 5.0. En caso de que se ingresen más de cinco notas para un estudiante, el programa deberá considerar únicamente las cinco notas más altas para los cálculos posteriores.

A partir de las notas admitidas, una vez sean validadas, el programa debe calcular: la nota más alta, la más baja, y el promedio de las notas consideradas.

Además, el programa debe calcular una nota definitiva, la cual se obtiene sumando al promedio una bonificación por cantidad de notas ingresadas. Este bono se asigna de la siguiente manera: 
- Si el estudiante registra únicamente el número mínimo de notas requeridas, no recibe bonificación
- Por cada nota adicional al mínimo (tres), hasta un máximo de cinco notas consideradas, se suma un bono fijo de 0.05 al promedio
- Si el estudiante registra más de cinco notas, el bono máximo aplicable será el correspondiente a cinco notas.

La nota mínima para aprobar es 3. Con base en el promedio, el programa debe indicar si el estudiante aprueba (promedio mayor o igual a 3) o reprueba (promedio menor que 3).

Finalmente, toda la información generada debe almacenarse en un dataframe de Pandas denominado ***notas_definitivas*** y mostrarse en pantalla.

In [1]:
import pandas as pd

#Definición de las cantidades máxima y mínima de notas por estudiante y de la bonificación por nota adicional
MIN_NOTAS = 3
MAX_NOTAS = 5
BONO_POR_NOTA = 0.05

nombres = []
nota_maxima = []
nota_minima = []
promedios = []
nota_definitiva = []
bonificacion = []
estado = []
cantidades_nota = [] #Las listas que contendrán los datos; una por cada columna del dataframe a crear

print("☰ GESTOR DE NOTAS ESTUDIANTILES ☰")
print(f"Debe ingresar, al menos, {MIN_NOTAS} notas por estudiante")

while True:
    try:
        cant_estudiantes = int(input("Cantidad de estudiantes a registrar: "))
        if cant_estudiantes > 0:
            break
        else:
            print("Debe ingresar al menos un estudiante") # Comprueba que la cantidad de est. no quede en cero
    except ValueError:
        print("Ingrese un número válido") # Comprueba que solo se ingresen números

for e in range(cant_estudiantes): # Ciclo for para registrar las notas para cada estudiante
    
    while True:
        nombre = input("Nombre del estudiante: ").strip()
        if nombre == "":
            print("El nombre no puede estar en blanco")
        elif any(char.isdigit() for char in nombre): # Valida que el nombre no contenga números
            print("El nombre no puede contener números")
        else:
            nombres.append(nombre)
            break

    while True:
        try:
            cant_notas = int(input(f"Cantidad de notas para {nombre}: "))
            if cant_notas >= MIN_NOTAS: # Comprueba que la cantidad de notas sea igual o mayor que tres
                cantidades_nota.append(cant_notas)
                break
            else:
                print(f"Recuerde ingresar al menos {MIN_NOTAS} notas")
        except ValueError:
            print("Ingrese un número válido") # Comprueba que solo se ingresen números
    print(f"Estudiante No. {e + 1}: ", nombre, " - ", cant_notas, "notas")

    notas = [] # Lista para aglutinarlas notas por cada estudiante
    n = 0
    
    while n < cant_notas:
        try:
            nota = float(input(f"Ingrese la nota No. {n + 1}: "))
            if 1.0 <= nota <= 5.0: # Comprueba que las notas ingresadas estén dentro del rango permitido
                notas.append(nota)
                n += 1
            else:
                print("La nota debe estar entre 1.0 y 5.0.")
        except ValueError:
            print("Ingrese un valor numérico válido.")
    
    if len(notas) > MAX_NOTAS: # Tiene en cuenta solo las cinco notas más altas, si hay más de cinco
        notas.sort(reverse=True) # Como sort() organiza de forma ascendente, debe reversarse para tomar las notas cinco más altas
        notas = notas[:MAX_NOTAS]

    promedio = round(sum(notas) / len(notas), 2)

    notas_efectivas = min(cant_notas, MAX_NOTAS)
    bono = (notas_efectivas - MIN_NOTAS) * BONO_POR_NOTA
    bono = max(0, bono)
    nota_final = round(promedio + bono, 2) #La nota definitiva es una métrica derivada del promedio y la bonificación

    nota_maxima.append(max(notas))# Los datos entran a las listas definidas al inicio
    nota_minima.append(min(notas))
    promedios.append(promedio) 
    bonificacion.append(bono)
    nota_definitiva.append(nota_final)
    estado.append("Aprobado" if nota_final >= 3 else "Reprobado") # Se asigna el estado dependiendo de la nota def.

#Aquí termina el ciclo for

notas_definitivas = pd.DataFrame({ # Se crea el dataFrame, mediante un diccionario, con las listas como columnas, 
    "Nombre": nombres,
    "Nota más Alta": nota_maxima,
    "Nota más Baja": nota_minima,
    "Cantidad de Notas" : cantidades_nota,
    "Promedio": promedios,
    "Bonificación": bonificacion,
    "Nota Definitiva": nota_definitiva,
    "Estado": estado
})

print("\nDataframe 'notas_definitivas':")
notas_definitivas

☰ GESTOR DE NOTAS ESTUDIANTILES ☰
Debe ingresar, al menos, 3 notas por estudiante
Estudiante No. 1:  Néstor  -  5 notas
Estudiante No. 2:  Jimena  -  4 notas
Estudiante No. 3:  Pedro  -  6 notas
Estudiante No. 4:  Idalia  -  3 notas
Estudiante No. 5:  Vicente  -  4 notas
Estudiante No. 6:  Olga  -  4 notas
Estudiante No. 7:  Carolina  -  4 notas
Estudiante No. 8:  Tránsito  -  7 notas
Recuerde ingresar al menos 3 notas
Estudiante No. 9:  Miguel  -  3 notas
Estudiante No. 10:  Joaquín  -  5 notas
Estudiante No. 11:  Isabel  -  3 notas
Estudiante No. 12:  Horacio  -  4 notas

Dataframe 'notas_definitivas':


,Nombre,Nota más Alta,Nota más Baja,Cantidad de Notas,Promedio,Bonificación,Nota Definitiva,Estado
0,Néstor,3.3,2.5,5,2.88,0.10,2.98,Reprobado
1,Jimena,4.0,2.6,4,3.23,0.05,3.28,Aprobado
2,Pedro,3.4,2.5,6,2.86,0.10,2.96,Reprobado
3,Idalia,3.5,2.8,3,3.10,0.00,3.10,Aprobado
4,Vicente,3.7,2.5,4,2.88,0.05,2.93,Reprobado
5,Olga,3.5,2.0,4,2.85,0.05,2.90,Reprobado
6,Carolina,3.5,1.0,4,2.52,0.05,2.57,Reprobado
7,Tránsito,4.0,3.0,7,3.66,0.10,3.76,Aprobado
8,Miguel,3.2,1.0,3,2.30,0.00,2.30,Reprobado
9,Joaquín,4.2,3.0,5,3.46,0.10,3.56,Aprobado


---

## **2.** Práctica Dos: Consultas a un DataFrame y su respectiva Equivalencia a SQL

Una vez creado el dataframe *notas_definitivas*, es posible llevar a cabo operaciones típicas de análisis de datos como emplear la función *describe()* para generar un perfil estadístico-descriptivo de columnas con valores cuantitativos.

In [3]:
notas_definitivas[['Cantidad de Notas', 'Promedio', 'Nota Definitiva']].describe()

,Cantidad de Notas,Promedio,Nota Definitiva
count,12.000000,12.00000,12.000000
mean,4.333333,3.00500,3.059167
std,1.230915,0.41082,0.430886
min,3.000000,2.30000,2.300000
25%,3.750000,2.83000,2.867500
50%,4.000000,2.88000,2.970000
75%,5.000000,3.28750,3.350000
max,7.000000,3.66000,3.760000


A partir del perfil estadístico-descriptivo anterior, por ejemplo, se evidencia que la media de cantidad de notas es **4.3**, superior a la cantidad mínima de notas admitible, que es 3. La media de la nota definitiva es **3.06**, ligeramente superior a la de la media de la columna *Promedio*, donde se muestran las notas sin bonificaciones; esto sugiere que las bonificaciones por tener más de tres notas no tuvieron mucho peso al definir el estado de aprobación o reprobación.

Además, la media de las notas definitivas logró superar el umbral de 3.0, la nota mínima para aprobar, si bien la distancia entre ambos valores es muy estrecha; el desempeño, a nivel general, fue bueno.

Sin embargo, dado que el tamaño muestral es solo de **12**, toda conclusión se debe tomar con cautela.

#### **2.1** Consulta que arroje la cantidad de estudiantes aprobados y reprobados

In [4]:
notas_definitivas.groupby("Estado").size()

Estado
Aprobado     5
Reprobado    7
dtype: int64

##### Equivalencia en SQL:

    SELECT estado, COUNT(*)
    FROM notas_definitivas
    GROUP BY estado;

#### **2.2** Consulta que arroje el promedio de la nota definitiva por estudiantes aprobados y reprobados

In [5]:
notas_definitivas.groupby("Estado")["Nota Definitiva"].mean()

Estado
Aprobado     3.460000
Reprobado    2.772857
Name: Nota Definitiva, dtype: float64

La consulta anterior, en SQL, tranformando los nombres de las columna y dataframe en denominaciones aceptables,
vendría siendo algo así:

    SELECT estado, AVG(nota_definitiva)
    FROM notas_definitivas
    GROUP BY estado;

#### **2.3** Consulta que arroje los valores máximo y mínimo de la nota más baja entre los estudiantes aprobados y reprobados

In [13]:
notas_definitivas.groupby("Estado")["Nota más Baja"].agg(["max", "min"])

,max,min
Estado,,
Aprobado,3.0,2.6
Reprobado,2.5,1.0


##### Equivalencia en SQL:

    SELECT estado, MAX(nota_mas_baja), MIN(nota_mas_baja)
    FROM notas_definitivas
    GROUP BY estado;

#### **2.4** Consulta que arroje la cantidad de estudiantes y el promedio de la nota definitiva por cantidad de notas presentadas

In [12]:
notas_definitivas.groupby("Cantidad de Notas")["Nota Definitiva"].agg(["count", "mean"])

,count,mean
Cantidad de Notas,,
3,3,2.723333
4,5,3.056000
5,2,3.270000
6,1,2.960000
7,1,3.760000


##### Equivalencia en SQL:

    SELECT cantidad_de_notas, COUNT(*) AS count, AVG(nota_definitiva) AS mean
    FROM notas_definitivas
    GROUP BY cantidad_de_notas;

#### **2.5** Consulta que arroje los registros completos de los tres estudiantes con la nota definitiva más alta, en orden descendente

In [11]:
notas_definitivas.sort_values("Nota Definitiva", ascending=False).head(3)

,Nombre,Nota más Alta,Nota más Baja,Cantidad de Notas,Promedio,Bonificación,Nota Definitiva,Estado
7,Tránsito,4.0,3.0,7,3.66,0.10,3.76,Aprobado
11,Horacio,4.0,2.7,4,3.55,0.05,3.60,Aprobado
9,Joaquín,4.2,3.0,5,3.46,0.10,3.56,Aprobado


##### Equivalencia en SQL:

    SELECT *
    FROM notas_definitivas
    ORDER BY nota_definitiva DESC
    LIMIT 3;

#### **2.6** Consulta que arroje el nombre, el estado y la nota definitiva de los tres estudiantes con la nota definitiva más baja, en orden ascendente

In [10]:
notas_definitivas[["Nombre", "Estado", "Nota Definitiva"]].nsmallest(2, "Nota Definitiva")

,Nombre,Estado,Nota Definitiva
8,Miguel,Reprobado,2.30
6,Carolina,Reprobado,2.57


##### Equivalencia en SQL:

    SELECT nombre, estado, nota_definitiva
    FROM notas_definitivas
    ORDER BY nota_definitiva
    LIMIT 2;

---

## **3.** Práctica Tres: Programación Orientada a Objetos

Desarrollar un programa que permita ingresar las medidas de los lados de varios triángulos. El programa debe validar que las medidas formen un triángulo correcto, es decir, que la suma de dos lados debe ser mayor que el tercero.

Utilizando una clase ***Triangulo***, determinar si cada triángulo es equilátero, isósceles o escaleno, según la medida de sus lados.

El programa debe solicitar cuántos triángulos se desean ingresar, mostrar la clasificación de cada uno y, al finalizar, presentar un conteo total de triángulos por tipo.

In [14]:
class Triangulo: # Se crea una clase para gestionar el registro del triángulo
    def __init__(self, lado1, lado2, lado3):
        self.lado1 = lado1
        self.lado2 = lado2
        self.lado3 = lado3

    def validez(self): #Método de la clase para validar que el triángulo sea matemáticamente correcto
        return (
            self.lado1 > 0 and self.lado2 > 0 and self.lado3 > 0 and
            self.lado1 + self.lado2 > self.lado3 and
            self.lado1 + self.lado3 > self.lado2 and
            self.lado2 + self.lado3 > self.lado1
        )

    def tipo(self): #Método de la clase para identificar el tipo de triángulo según las medidas de sus lados
        #Por medio de un set{}, y de len(), se determina cuántos lados están repetidos
        tres_lados = len({self.lado1, self.lado2, self.lado3})
        if tres_lados == 1:
            return "Equilátero"
        elif tres_lados == 2:
            return "Isósceles"
        else:
            return "Escaleno"

#Función para gestionar que las medidas de los lados sean valores aceptables
def ingreso_lado(mensaje):
    while True:
        try:
            valor = float(input(mensaje))
            if valor > 0:
                return valor
        except ValueError:
            print("Debe ingresar un número válido")


print("☰ CLASIFICADOR Y CONTADOR DE TRÍANGULOS ◭ ☰")

while True:
    try:
        cant_tg = int(input("¿Cuántos triángulos ingresará?: "))
        if cant_tg > 0:
            break
        print("Debe ingresar al menos un triángulo")
    except ValueError:
        print("Ingrese un número válido")

contador = {
    "Equilátero": 0,
    "Isósceles": 0,
    "Escaleno": 0
}

for i in range(cant_tg):
    while True:
        l1 = ingreso_lado("Medida del 1er lado: ")
        l2 = ingreso_lado("Medida del 2do lado: ")
        l3 = ingreso_lado("Medida del 3er lado: ")
        triangulo = Triangulo(l1, l2, l3)

        if triangulo.validez():
            break
        else:
            print("Triángulo no válido")

    tipo = triangulo.tipo()
    print(f"Triángulo No. {i + 1} / Lados: {l1} - {l2} - {l3} / Es un triángulo", tipo)
    contador[tipo] += 1


print("\nDistribución de Triángulos Ingresados")

print("Triángulos Equiláteros:", contador["Equilátero"])
print("Triángulos Isósceles:", contador["Isósceles"])
print("Triángulos Escalenos:", contador["Escaleno"])

☰ CLASIFICADOR Y CONTADOR DE TRÍANGULOS ◭ ☰
Triángulo No. 1 / Lados: 2.5 - 2.8 - 3.1 / Es un triángulo Escaleno
Triángulo No. 2 / Lados: 6.0 - 5.0 - 5.0 / Es un triángulo Isósceles
Triángulo no válido
Triángulo No. 3 / Lados: 2.5 - 3.0 - 3.4 / Es un triángulo Escaleno
Triángulo No. 4 / Lados: 5.0 - 5.0 - 5.0 / Es un triángulo Equilátero
Triángulo No. 5 / Lados: 8.0 - 6.7 - 6.7 / Es un triángulo Isósceles
Triángulo No. 6 / Lados: 0.8 - 0.8 - 0.8 / Es un triángulo Equilátero
Triángulo No. 7 / Lados: 9.0 - 8.0 - 8.0 / Es un triángulo Isósceles
Triángulo No. 8 / Lados: 3.5 - 4.12 - 4.1 / Es un triángulo Escaleno
Triángulo No. 9 / Lados: 5.0 - 5.8 - 5.0 / Es un triángulo Isósceles

Distribución de Triángulos Ingresados
Triángulos Equiláteros: 2
Triángulos Isósceles: 4
Triángulos Escalenos: 3
